File: Graph/Floyd-WarshallAlgorithm/Floyd-WarshallAlgorithm.cpp
#include <iostream>
#include <vector>
#include <limits>

const int INF = std::numeric_limits<int>::max();

std::vector<std::vector<int>> floydWarshall(const std::vector<std::vector<int>>& graph) {
    int V = graph.size();
    std::vector<std::vector<int>> dist(V, std::vector<int>(V, INF));

    for (int i = 0; i < V; ++i) {
        for (int j = 0; j < V; ++j) {
            if (i == j) {
                dist[i][j] = 0;
            } else if (graph[i][j] != INF) {
                dist[i][j] = graph[i][j];
            }
        }
    }

    for (int k = 0; k < V; ++k) {
        for (int i = 0; i < V; ++i) {
            for (int j = 0; j < V; ++j) {
                if (dist[i][k] != INF && dist[k][j] != INF && dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    return dist;
}

int main() {
    std::vector<std::vector<int>> graph = {
        {0, 3, INF, 7},
        {8, 0, 2, INF},
        {5, INF, 0, 1},
        {2, INF, INF, 0}
    };

    std::vector<std::vector<int>> result = floydWarshall(graph);

    std::cout << "Matrix of shortest paths:\n";
    for (const auto& row : result) {
        for (int value : row) {
            if (value == INF) {
                std::cout << "INF ";
            } else {
                std::cout << value << " ";
            }
        }
        std::cout << "\n";
    }

    return 0;
}

File: Graph/Floyd-WarshallAlgorithm/Floyd-WarshallAlgorithm.py
INF = float('inf')

def floyd_warshall(graph):
    V = len(graph)
    dist = [[INF] * V for _ in range(V)]

    for i in range(V):
        for j in range(V):
            if i == j:
                dist[i][j] = 0
            elif graph[i][j] is not None:
                dist[i][j] = graph[i][j]

    for k in range(V):
        for i in range(V):
            for j in range(V):
                if dist[i][k] != INF and dist[k][j] != INF and dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]

    return dist


# Пример использования
graph = [
    [0, 3, None, 7],
    [8, 0, 2, None],
    [5, None, 0, 1],
    [2, None, None, 0]
]

result = floyd_warshall(graph)
print("Матрица кратчайших путей:")
for row in result:
    print(row)